client/src/worker_registry/mod.rs

```diff
@@ -0,0 +1,288 @@
+//! This module enables the tracking of workers that are associated with a client instance.
+//! This is needed to implement Eager Workflow Start, a latency optimization in which the client,
+//!  after reserving a slot, directly forwards a WFT to a local worker.
+
+use rand::{seq::SliceRandom, thread_rng};
+use std::{collections::HashMap, fmt::Debug};
+use temporal_sdk_core_protos::temporal::api::workflowservice::v1::PollWorkflowTaskQueueResponse;
+
+/// This trait wraps a worker, which provides WFT processing slots.
+#[cfg_attr(test, mockall::automock)]
+pub trait SlotProvider: Send + Sync + Debug {
+    /// A unique identifier for the worker.
+    fn uuid(&self) -> String;
+    /// The namespace for the WFTs that it can process.
+    fn namespace(&self) -> String;
+    /// The task queue this provider listens to.
+    fn task_queue(&self) -> String;
+    /// Try to reserve a slot with matching namespace and task queue.
+    fn try_reserve_wft_slot(&self) -> Option<Box<dyn Slot>>;
+}
+
+/// This trait represents a slot reserved for processing a WFT by a worker.
+#[cfg_attr(test, mockall::automock)]
+pub trait Slot: Send + Sync {
+    /// Consumes this slot by dispatching a WFT to its worker. This can only be called once.
+    fn schedule_wft(&mut self, task: PollWorkflowTaskQueueResponse) -> Result<(), anyhow::Error>;
+}
+
+/// This trait enables local workers to made themselves visible to a shared client instance.
+pub trait WorkerRegistry: Send + Sync {
```

Comment:
Could also have a single `register` that returns a `WorkerPermit` with `unregister` on it. That way you could actually get rid of this trait and change `fn workers(&self) -> Arc<RwLock<dyn WorkerRegistry>> {` to something like `fn register_slot_provider(&self, Box<dyn SlotProvider>) -> WorkerPermit`. Also it will simplify your unregister logic instead of asking the caller to give you their ID again. Also may be worth choosing to use either "worker" or "slot provider" terminology instead of both here.

(none of my suggestions like these are required, just discussion starters)

Reply:
With the move to a single worker per namespace+queue+client it helps to have a worker id on register to detect multiple workers claiming the same namespace+queue+client, and ignore the registration. The logic on the caller worker gets easier too, they don't have to deal with whether they got a permit or not. So I think I'll stick to register(provider)/unregister(id)...

Reply:
Can't you just consider a second attempted registration of the same namespace+queue a failure? Why do you have to account for identifier at all?

Reply:
No, I can't,  because registration will happen whether the programmer is using eager start or not in their code, and I want to make sure that I don't break anything in existing code... I also worry about a worker trying to register twice, and the id helps to decide to do a warn! or silently ignore it.

Reply:
I think this kind of registry/registration should be eager-workflow-start specific and don't perform eager-workflow-enabled-worker-register if that's disabled.

Reply:
The goal is that eager will always be enabled, and it is up to the app to add a flag to start_workflow to use eager. Since we don't know that beforehand , we have to assume that they may use eager, and we need to always register...

Reply:
I don't think that's the goal though, and at least for now it's always disabled. Eager workflow start capability is enabled/disabled at a worker level (but we may allow additional overriding to disable it at the start level if enabled at the worker level) and it will be disabled by default (and we're by no means settled on whether that will change in the future).

Reply:
I don't think it is enabled/disabled at worker level in neither java or go.  It is also an unnecessary complexity. Once we decide that we want to enable eager in the server, it should just work if clients use the flag in start_workflow....

Reply:
I see, my mistake! So I guess you will register every one as eager-capable.

So how would a worker ever call register on itself twice necessitating an identifier (instead of once and given a thing for it to unregister itself)?

Reply:
With the current implementation it won't happen, if the constructor fails you get a new id. But I was thinking that one day the registration may not be just for single process, but for a pool of local processes that can all benefit from early start, and having  a unique id for a worker in the trait method will hide that implementation. For debugging it also helps to have a unique id per worker, specially if we are imposing the restriction that only one can register.
 
And as I just mentioned, the "given a thing to unregister" forces the client to deal with failed registrations, while unregister with id is just transparent.

Reply:
I personally wouldn't be concerned with a potential future of deduplicating, and having an optional "lease" for a successful registration stored on the worker to unregister is normal. The obvious benefits are 1) you don't need a separate trait compared to a single call, and 2) a worker doesn't have to be responsible for its identifier. But if consensus is that we want workers to have this identifier, ok.